elastic/beats
# [Heartbeat] Handle data streams

Handles data streams from fleet, the current heartbeat code doesn't handle `data_streams` at all, this fixes that. Additionally, it hoists the `id` from the top level of the yaml config, and merges both levels of `data_streams`, since one is at the input level, and the other is at the stream level.

Additionally, this cleans up the data streams code significantly, introducing a new `add_data_streams_index` processor that:
1. More efficiently formats index names for data streams
2. Allows individual events to override the dataset (useful for synthetics where we have a base `browser` dataset, but also `browser_screenshot` and `browser_network` for extended data that can take up lots of space, and often requires a different ILM policy).

## Checklist



~~- [ ] I have made corresponding changes to the documentation~~
~~- [ ] I have made corresponding change to the default configuration files~~

## How to test this PR locally

Use the following `heartbeat.yml` and verify the output sets the correct monitor ID and writes to the correct data streams

```yaml
# Configure monitors inline
heartbeat.monitors:
- id: 4ae879a9-b5da-4132-94ba-ab6e1fcbdc6e
  name: Sample monitor 2
  revision: 1
  type: synthetics/http
  use_output: default
  meta:
    package:
    name: synthetics
    version: 0.1.28
  data_stream:
    namespace: default
  streams:
   - id: synthetics/http-http-4ae879a9-b5da-4132-94ba-ab6e1fcbdc6e
     name: Sample monitor 2
     type: http
     data_stream:
       dataset: http
       type: synthetics
     urls: 'http://elastic.co'
     service.name: APM Service Name
     schedule: '@every 5s'
     timeout: 1600
     max_redirects: 1
     proxy_url: 'http://elastic.co'
     tags:
       - tag
       - tag2 
```
